ccm_web/client/src/pages/BulkSectionCreate.tsx

```diff
@@ -0,0 +1,534 @@
+
+    if (sectionNames.length === 1 && isHeader(sectionNames[0])) {
+      invalidations.push({ error: 'No data', type: InvalidationType.Error })
+    }
+    return invalidations
+  }
+}
+
+// For validating row level issues
+interface SectionsRowInvalidation {
+  message: string
+  rowNumber: number
+  type: InvalidationType
+}
+
+interface SectionRowsValidator {
+  validate: (sectionName: string[]) => SectionsRowInvalidation[]
+}
+
+class DuplicateSectionInFileSectionRowsValidator implements SectionRowsValidator {
+  validate = (sectionNames: string[]): SectionsRowInvalidation[] => {
+    const sortedSectionNames = sectionNames.map(n => { return n.toUpperCase() }).sort((a, b) => { return a.localeCompare(b) })
+    const duplicates: string[] = []
+    for (let i = 1; i < sortedSectionNames.length; ++i) {
+      if (sortedSectionNames[i - 1] === sortedSectionNames[i] && !duplicates.includes(sortedSectionNames[i])) {
+        duplicates.push(sortedSectionNames[i])
+      }
+    }
+    if (duplicates.length === 0) {
+      return []
+    }
+    const invalidations: SectionsRowInvalidation[] = []
+    let i = 1
+    sectionNames.forEach(sectionName => {
+      if (duplicates.includes(sectionName.toUpperCase())) {
+        invalidations.push({ message: 'Duplicate section name: "' + sectionName + '"', rowNumber: i + 1, type: InvalidationType.Error })
+      }
+      ++i
+    })
+    return invalidations
+  }
+}
+
+interface BulkSectionCreateProps extends CCMComponentProps {
+
+}
+
+function BulkSectionCreate (props: BulkSectionCreateProps): JSX.Element {
+  const classes = useStyles()
+  const confirmationClasses = useConfirmationStyles()
+  const rowLevelErrorClasses = useRowLevelErrorStyles()
+  const topLevelClasses = useTopLevelErrorStyles()
+  const apiErrorClasses = useAPIErrorStyles()
+
+  const [isExistingSectionsLoading, setIsExistingSectionsLoading] = useState(true)
+  const [pageState, setPageState] = useState<BulkSectionCreatePageStateData>({ state: BulkSectionCreatePageState.LoadingExistingSectionNames, schemaInvalidation: [], rowInvalidations: [] })
+  const [file, setFile] = useState<File|undefined>(undefined)
+  const [sectionNames, setSectionNames] = useState<string[]>([])
+  const [existingSectionNames, setExistingSectionNames] = useState<string[]|undefined>(undefined)
+
+  const loadCanvasSectionData = (): void => {
+    getCourseSections(props.ltiKey, 'TODO-CourseNumberFromProps?')
+      .then(sections => {
+        setExistingSectionNames(sections.map(s => { return s.toUpperCase() }))
+        setIsExistingSectionsLoading(false)
+      }).catch(() => {
+        setExistingSectionNames(undefined)
+        setPageState({ state: BulkSectionCreatePageState.LoadingExistingSectionNamesFailed, schemaInvalidation: [], rowInvalidations: [] })
+      })
+  }
+
+  // Support for this might be addressed by https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/74 ?
+  useEffect(() => {
+    loadCanvasSectionData()
+  }, [])
```

Comment:
I would ideally start the loadCanvasSectionData call on page load w/o blocking the file upload component, and only show the waiting  overlay if the call hasn't completed yet, however this was simpler to implement, and if they are coming to this feature odds are they are going to upload a file so we will need to make the call.. might as well make the call right away?

Could argue either way I guess.  If it ends up being an expensive call it would make sense to wait until we know we need to do it.

Reply:
Let's wait to see what happens once the API call triggers a communication with Canvas.